File: controllersProyect/Productividad/BackPressure.py
import math as mt


class BackPressure:
    def __init__(self):
        print("Initializing")

    def defaultMethod(self, pr, pwf1, pwf2, qg1, qg2, r):
        a = mt.log(pow(pr, 2) - pow(pwf2, 2), 10) - mt.log(
            pow(pr, 2) - pow(pwf1, 2), 10
        )
        b = mt.log(qg2, 10) - mt.log(qg1, 10)
        
        # Calcular m
        m = a / b  # Esta es m

        # Calculo de n
        n = 1 / m

        # Calculo de C
        c = qg1 / pow(pow(pr, 2) - pow(pwf1, 2), n)
        
        # Calculo de qg
        pwf = [pr]
        for i in range(1, r + 1):
            aux = round((pr - 14.7) / r, 3)
            pwf.append(round(pr - (aux * i), 3))  # Metodo para agregar datos de pwf
            if i == r:
                pwf.append(0)
        qg = list()
        aux2 = 0
        for i in pwf:
            qg.append(round(c * pow((pow(pr, 2) - pow(i, 2)), n), 3))
            pwf[aux2] = round(pow(pr, 2) - pow(pwf[aux2], 2), 3)
            aux2 += 1

        return {"m": m, "n": n, "c": c, "lpwf": pwf, "lqg": qg}

File: assets/js/BackPressure.js
document.addEventListener("DOMContentLoaded", function () {
  const nav = document.querySelector("#nav");
  const abrir = document.querySelector("#abrir");
  const cerrar = document.querySelector("#cerrar");

  abrir.addEventListener("click", () => {
    nav.classList.add("visible");
  });

  cerrar.addEventListener("click", () => {
    nav.classList.remove("visible");
  });

  /**
   * Metodo para graficas backpressure
   */
  const chartContext1 = document.getElementById("myChart").getContext("2d");
  const chartContextModal1 = document
    .getElementById("modalChart1")
    .getContext("2d");
  // Valores iniciales
  const initialData = {
    labels: [10, 20, 30, 20, 10],
    datasets: [
      {
        label: "Vogel",
        backgroundColor: "rgba(75, 192, 192, 0.2)",
        borderColor: "rgba(75, 192, 192, 1)",
        borderWidth: 1,
        data: [10, 20, 30, 20, 10],
        fill: false,
        tension: 0.4,
      },
    ],
  };
  const chartOptions = {
    scales: {
      x: {
        type: "logarithmic",
        title: {
          display: true,
          text: "Qo (Logarítmico)",
        },
      },
      y: {
        type: "logarithmic",
        title: {
          display: true,
          text: "Pwf (Logarítmico)",
        },
        beginAtZero: true,
      },
    },
  };
  var myChart = new Chart(chartContext1, {
    type: "line", // Cambiado a tipo de gráfico de línea para mejorar la visualización logarítmica
    data: initialData,
    options: chartOptions,
  });
  var myChartModal = new Chart(chartContextModal1, {
    type: "line", // Cambiado a tipo de gráfico de línea para mejorar la visualización logarítmica
    data: initialData,
    options: chartOptions,
  });

  //Instanciar boton
  const enviar = document.getElementById("calcular");
  enviar.addEventListener("click", function () {
    dataEnv = {
      pr: document.getElementById("pws").value,
      pwf1: document.getElementById("pwf1").value,
      pwf2: document.getElementById("pwf2").value,
      qg1: document.getElementById("qg1").value,
      qg2: document.getElementById("qg2").value,
      r: document.getElementById("r").value,
    };

    $.ajax({
      type: "POST",
      url: "/productividad/back-pressure",
      data: JSON.stringify(dataEnv),
      contentType: "application/json",
      dataType: "JSON",
      success: done,
      error: errorGraphic,
    });

    function done(result) {
      console.log(result);
      imprimirResultados(result);

      const newData = {
        labels: result.lqg,
        datasets: [
          {
            label: "BackPressure",
            backgroundColor: "rgba(75, 192, 192, 0.2)",
            borderColor: "rgba(75, 192, 192, 1)",
            borderWidth: 1,
            data: result.lpwf,
            tension: 1,
          },
        ],
      };
      myChart.data = newData;
      myChart.update();
      myChartModal.data = newData;
      myChartModal.update();
    }
    function errorGraphic(error) {
      alert(error.message);
    }

    function imprimirResultados(data) {
      const contenedor = document.getElementById("results");
      contenedor.innerHTML = ""; // Limpiar contenido anterior
      const divResultado = document.createElement("div");

      $.each(data, function (key, value) {
        const pElement = document.createElement("p");
        pElement.textContent = `${key}: ${JSON.stringify(value)}`;
        divResultado.appendChild(pElement);
      });

      contenedor.appendChild(divResultado);
    }
  });
});
